libs/walproposer/src/walproposer.rs

```diff
@@ -17,7 +17,7 @@ use crate::{
 /// Refer to `pgxn/neon/walproposer.h` for documentation.
 pub trait ApiImpl {
     fn get_shmem_state(&self) -> *mut crate::bindings::WalproposerShmemState {
-        todo!()
+        unsafe { crate::bindings::GetWalpropShmemState() }
```

Comment:
Already written in another comment, it won't work for multiple walproposers running at the same time. But you can make `ApiImpl` implementations maintain their own "shmem" if you need to

Reply:
I would actually like to not use thread_local as well, because it would mean that you can't send `struct WalProposer` to another thread and use it from there.

What I want to see is `MockImpl` implementing `get_shmem_state`, and `pub trait ApiImpl {` doesn't contain any implementations.
Look at `fn get_shmem_state` at `walproposer_api.rs:584`, it uses `UnsafeCell` to store shmem content. In `MockImpl` you can do the same.

Reply:
Done